File: spring_java/src/main/java/dev/pollito/spring_java/config/security/handler/AuthenticationErrorResponseWriter.java
package dev.pollito.spring_java.config.security.handler;

import static io.opentelemetry.api.trace.Span.current;
import static java.time.OffsetDateTime.now;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import org.jspecify.annotations.NonNull;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;

final class AuthenticationErrorResponseWriter {

  private AuthenticationErrorResponseWriter() {}

  static void write(
      @NonNull HttpServletRequest request,
      @NonNull HttpServletResponse response,
      @NonNull HttpStatus status,
      @NonNull String detail)
      throws IOException {
    response.setStatus(status.value());
    response.setContentType(MediaType.APPLICATION_JSON_VALUE);

    String json =
        "{"
            + "\"title\":\""
            + escapeJson(status.getReasonPhrase())
            + "\","
            + "\"detail\":\""
            + escapeJson(detail)
            + "\","
            + "\"instance\":\""
            + escapeJson(request.getRequestURI())
            + "\","
            + "\"status\":"
            + status.value()
            + ","
            + "\"timestamp\":\""
            + now()
            + "\","
            + "\"trace\":\""
            + current().getSpanContext().getTraceId()
            + "\""
            + "}";

    response.getWriter().print(json);
  }

  private static @NonNull String escapeJson(@NonNull String value) {
    return value.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n");
  }
}

File: spring_kotlin/src/main/kotlin/dev/pollito/spring_kotlin/config/security/handler/AuthenticationErrorResponseWriter.kt
package dev.pollito.spring_kotlin.config.security.handler

import io.opentelemetry.api.trace.Span
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import java.time.OffsetDateTime
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType.APPLICATION_JSON_VALUE

object AuthenticationErrorResponseWriter {

  fun write(
      request: HttpServletRequest,
      response: HttpServletResponse,
      status: HttpStatus,
      detail: String,
  ) {
    response.status = status.value()
    response.contentType = APPLICATION_JSON_VALUE

    val json =
        "{" +
            "\"title\":\"${escapeJson(status.reasonPhrase)}\"," +
            "\"detail\":\"${escapeJson(detail)}\"," +
            "\"instance\":\"${escapeJson(request.requestURI)}\"," +
            "\"status\":${status.value()}," +
            "\"timestamp\":\"${OffsetDateTime.now()}\"," +
            "\"trace\":\"${Span.current().spanContext.traceId}\"" +
            "}"

    response.writer.print(json)
  }

  private fun escapeJson(value: String): String {
    return value.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n")
  }
}
